### PR TITLE
tinystdio: fgets returns NULL only when EOF is reached

### PR DESCRIPTION
**Pull request by HanaMAshour**
According to the ISO/IEC_9899_1999, section:7.19.7.2 It is mentioned that when EOF is encountered and no characters were read into the array a null pointer is returned. In the fgets function in picolibc,
it always returned NULL when it reaches EOF even if characters were read and that the file is not empty.

Accordingly a flag was added that checks if the file is empty or not, there is an added "if" condition that checks on this flag, if EOF was encountered for whether NULL or the pointer to the array that was read will be returned. The case where it returns NULL occurs when EOF and no characters were read. This change checks and handles that fgets returns the pointer to the array correctly when it reaches EOF.

This is a simple testcase that shows the effect of this change:

```c
#include <stdio.h>
#include <string.h>

const char *string = "123456789";    

int main(){

    char *pchar;
    char line[12];
    int compare_result;
    FILE *file;

    file = fopen("testfile.dat", "w");
    if(file == NULL) return 1;
    fputs(string, file);
    fclose(file);

    file = fopen( "testfile.dat", "r" );
    if(file == NULL) return 1;
    
    /*Calling fgets to reach EOF and check on the returned value*/
    pchar = fgets( line, 12, file);
    if(pchar == line)
        printf("\nfgets return is correct\n");
    else
        printf("\nfgets return is wrong!!\n");

    compare_result = strcmp( line, string );
    if(compare_result == 0)
        printf("\ncompare results passed\n");
    else
        printf("\ncompare results failed!!\n");
  
    fclose( file );

    return 0;
}
```

### CODE REVIEW

File: newlib/libc/tinystdio/fgets.c
Comment (keith-packard) [inline newlib/libc/tinystdio/fgets.c:47]: The `gets` code tests the pointer to see if it has moved. Can we do that here by using `if (cp == str)`? That would make the two versions a little more similar and avoid adding another variable to this code.

**Review by keith-packard — CHANGES_REQUESTED** [review]
I've taken your test case and added that on the hanaa_fgets_fix branch. Can you fix the copyright on that and add that to your branch before we merge this in?

File: test/test-fgets-eof.c
Comment (keith-packard) [inline test/test-fgets-eof.c:4]: Do you want to change the name used here?

Comment (HanaMAshour) [reply test/test-fgets-eof.c]: I changed the copyright and it is fixed now

File: newlib/libc/tinystdio/fgets.c
Comment (HanaMAshour) [reply newlib/libc/tinystdio/fgets.c]: I added the suggested change in the fgets function

**Comment by abrodkin** [comment]
@keith-packard requested changes are done, please consider merging.

**Comment by keith-packard** [comment]
rebased to make sure CI remains happy with this; will merge when tests are green

**Pull request merged**
